File: trpg-frontend/src/services/room.ts
import type { CreateRoomResult, ModuleSummary, MyRoomSummary, RoomPreview } from 'trpg-sdk';
import { useRoomStore } from '@/stores/room-store';
import { getAuthToken, sdk } from './api-client';

export type { CreateRoomResult, ModuleSummary, MyRoomSummary, RoomPreview };

// 房主/已加入玩家专属的操作（选模组/开始游戏/结束游戏/我的房间列表）需要
// 后端的房间重连凭证（X-Reconnect-Token，issue #39），加入/创建房间时签发、
// 存进 room-store——直接从 store 读，页面组件不需要在每次调用时手动传。
function requireReconnectToken(): string {
  const token = useRoomStore.getState().reconnectToken;
  if (!token) throw new Error('缺少房间重连凭证，请重新加入房间');
  return token;
}

// 创建/加入房间和「我的游戏」用的是**账号**凭证，不是上面那个房间凭证
// （issue #106）。两者分工：账号解决跨设备/跨时间找回，reconnectToken 解决
// 同一局进行中的快速重连。
function requireAuthToken(): string {
  const token = getAuthToken();
  if (!token) throw new Error('请先登录');
  return token;
}

// 创建房间（房主创建即加入，见 §5.2.5）
export async function createGameRoom(
  nickname?: string,
  roomName?: string,
  maxPlayers?: number
): Promise<CreateRoomResult> {
  return sdk.rooms.create(
    { nickname, roomName: roomName ?? '', maxPlayers: maxPlayers ?? 4 },
    requireAuthToken()
  );
}

// 拉取可用模组列表（本次没有做模组导入，只有一款内置模拟模组）
export async function listModules(): Promise<ModuleSummary[]> {
  return sdk.rooms.listModules();
}

// 房主确定模组
export async function selectModule(roomId: string, moduleId: string): Promise<void> {
  await sdk.rooms.selectModule(
    roomId,
    { moduleId, attributeGenMethod: 'point_buy' },
    requireReconnectToken()
  );
}

// 用房间码加入房间。issue #106 起后端按**账号**幂等：已经是这个房间的成员时
// 原样返回既有身份，所以这个函数同时承担「加入」和「掉线/换设备后重连」两个
// 用途——之前那句「已是本房间玩家则幂等返回已有身份」的注释是假的，后端当时
// 根本不检查，重复调用会给同一个人建重复玩家行。
export async function joinRoomByCode(
  roomCode: string,
  nickname?: string
): Promise<CreateRoomResult> {
  return sdk.rooms.join(roomCode, { nickname }, requireAuthToken());
}

// 获取房间信息（房间码预览）
export async function getRoomInfo(roomCode: string): Promise<RoomPreview> {
  return sdk.rooms.getInfo(roomCode);
}

// 房主点击「开始游戏」，从大厅推进到背景介绍——访客端轮询这个标记自动跟进
export async function startStory(roomId: string): Promise<void> {
  await sdk.rooms.startStory(roomId, requireReconnectToken());
}

// 我的房间列表——用于「浏览已有游戏」入口。issue #106 起按**账号**返回该用户
// 参与过的全部房间（换台设备登录同一账号也能看到），不再是「这个浏览器的最后
// 一个房间」。没登录时返回空列表而不是报错：这个入口在未登录状态下也会被渲染。
export async function listMyRooms(): Promise<MyRoomSummary[]> {
  const token = getAuthToken();
  if (!token) return [];
  return sdk.rooms.listMyRooms(token);
}

// 房主结束游戏，房间转入「已完成」状态，之后只能查看复盘
export async function endGame(roomId: string): Promise<void> {
  await sdk.rooms.endGame(roomId, requireReconnectToken());
}

File: trpg-backend/app/dto/room.py
"""Room 模块的 pydantic 请求/响应模型。

对应的 TS 类型由 trpg-sdk 的 codegen 脚本从这些模型生成（见 scripts/export_schema.py
和 issue #75），不再需要手动同步 trpg-sdk/src/types.ts。

命名约定：
- 后端代码内统一使用 snake_case Python 命名
- 通过 alias_generator 实现 JSON 层的 camelCase ↔ snake_case 自动映射
- 请求（camelCase JSON → snake_case Python）和响应（snake_case Python → camelCase JSON）
  由 pydantic 自动处理，业务代码无需关心
"""


from pydantic import Field, field_validator

from app.dto.common import CamelModel, UtcDatetime

# ── 请求体 ──────────────────────────────────────


class RoomCreate(CamelModel):
    """POST /api/v1/rooms 请求体"""

    nickname: str | None = Field(default=None, max_length=100)
    room_name: str = Field(..., min_length=1, max_length=200)
    max_players: int = Field(default=4, ge=1, le=20)

    @field_validator("nickname", "room_name")
    @classmethod
    def strip_text(cls, value: str | None) -> str | None:
        if value is None:
            return None
        stripped = value.strip()
        if not stripped:
            raise ValueError("不能为空")
        return stripped


class SelectModuleBody(CamelModel):
    """POST /api/v1/rooms/{roomId}/module 请求体"""

    module_id: str = Field(..., min_length=1)
    attribute_gen_method: str = Field(default="point_buy")


class JoinRoomBody(CamelModel):
    """POST /api/v1/rooms/{roomCode}/join 请求体"""

    nickname: str | None = Field(default=None, max_length=100)

    @field_validator("nickname")
    @classmethod
    def strip_nickname(cls, value: str | None) -> str | None:
        if value is None:
            return None
        stripped = value.strip()
        if not stripped:
            raise ValueError("昵称不能为空")
        return stripped


# ── 响应体 ──────────────────────────────────────


class RoomCreateResult(CamelModel):
    """POST /api/v1/rooms 返回"""

    room_id: str
    room_code: str
    reconnect_token: str
    player_id: str


class RoomPlayerRead(CamelModel):
    """房间内玩家摘要。

    注意 `player_id` 对应 ORM `Player` 的主键属性 `id`（名字不一样），所以不能直接
    `model_validate(player_orm)`——调用方需要显式映射 `player_id=p.id`（见
    service/room.py 的 _to_room_preview）。`from_attributes=True` 仍保留，方便
    其余名字一致的字段。camelCase 别名生成、populate_by_name 继承自 `CamelModel`——
    pydantic 的 `model_config` 在子类里是合并而非整体覆盖父类配置，这里不需要
    重复声明（issue #77 审计发现 #1，原先这里重写了一份和父类一样的配置，是
    #75 遗留的死代码）。
    """

    model_config = {"from_attributes": True}
    player_id: str
    nickname: str
    is_host: bool
    ready: bool
    has_character: bool


class ModuleRead(CamelModel):
    """模组信息（对应内容库 `Scenario` 表，`from_attributes=True` 支持直接从
    ORM 对象构造）。"""

    model_config = {"from_attributes": True}
    id: str
    title: str
    version: str
    authors: list[str]
    players_min: int
    players_max: int
    difficulty: int
    estimated_duration: str | None = None


class RoomPreview(CamelModel):
    """GET /api/v1/rooms/{roomCode} 返回"""

    room_id: str
    room_code: str
    room_name: str
    phase: str
    story_started: bool
    module_title: str | None = None
    player_count: int
    max_players: int
    players: list[RoomPlayerRead]


class MyRoomSummary(CamelModel):
    """GET /api/v1/me/rooms 返回项"""

    room_id: str
    room_code: str
    room_name: str
    phase: str
    module_title: str | None = None
    player_count: int
    max_players: int
    updated_at: UtcDatetime
